apiservers/engine/backends/image.go

```diff
@@ -49,7 +64,79 @@ func (i *Image) ImageHistory(imageName string) ([]*types.ImageHistory, error) {
 }
 
 func (i *Image) Images(filterArgs string, filter string, all bool) ([]*types.Image, error) {
-	return nil, fmt.Errorf("%s does not implement image.Images", i.ProductName)
+
+	result := []*types.Image{}
+
+	host, err := os.Hostname()
+	if err != nil {
+		return result,
+			derr.NewErrorWithStatusCode(fmt.Errorf("image.Images got unexpected error getting hostname"),
+				http.StatusInternalServerError)
+	}
+
+	params := &storage.ListImagesParams{
+		StoreName: host,
+	}
+
+	client := PortLayerClient()
+	if client == nil {
+		return result,
+			derr.NewErrorWithStatusCode(fmt.Errorf("image.Images failed to create a portlayer client"),
+				http.StatusInternalServerError)
+	}
+
+	images, err := client.Storage.ListImages(params)
+	if err != nil {
+		return result,
+			derr.NewErrorWithStatusCode(err, http.StatusInternalServerError)
+	}
+
+	// build a map from image id to image v1Compatibility metadata
+	v1Map := make(map[string]*v1.V1Image)
+	for _, image := range images.Payload {
+
+		if image.ID == "scratch" {
+			continue
+		}
+
+		v1Image := &v1.V1Image{}
+		decoder := json.NewDecoder(strings.NewReader(image.Metadata["v1Compatibility"]))
+		if err := decoder.Decode(v1Image); err != nil {
+			log.Fatal(err)
+		}
+		v1Map[image.ID] = v1Image
+	}
+
+	for id, v1Image := range v1Map {
+
+		size := getImageSize(v1Map, id)
```

Comment:
Yeah I was thinking about this. Please correct me if my thought process is wrong here:

Since we are sorting the resulting slice of images anyway by create time, and since a parent image always has an older create time than its child, we should be able to step through the sorted slice from oldest to newest and calculate each image's size in a single pass.

I've changed up the code to reflect this, what do you think?